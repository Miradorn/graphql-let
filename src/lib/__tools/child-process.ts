import execa, { ExecaChildProcess } from 'execa';
import terminate from 'terminate';

export async function killApp(app: ExecaChildProcess) {
  console.info(`Terminating app ${app.pid}...`);
  await new Promise((resolve, reject) => {
    terminate(app.pid, (err: any) => {
      if (err) return reject(err);
      return resolve(undefined);
    });
  });
  console.info(`App ${app.pid} was terminated`);
}

export function timeout(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export function spawn(
  command: string,
  args: string[],
  opts?: execa.CommonOptions<'utf-8'>,
) {
  return execa(command, args, {
    stdin: 'ignore',
    stdout: 'inherit',
    stderr: 'inherit',
    ...opts,
  });
}
