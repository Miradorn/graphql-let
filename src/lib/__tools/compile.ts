import memoryfs from 'memory-fs';
import path from 'path';
import webpack from 'webpack';
import nodeExternals from 'webpack-node-externals';
import type { GraphQLLetLoaderOptions } from '../../loader';

export default function compile(
  cwd: string,
  entries: string[],
  target: 'node' | 'web',
  gqlLetLoaderOptions: GraphQLLetLoaderOptions | undefined = undefined,
): Promise<webpack.Stats> {
  const compiler = webpack({
    mode: 'production',
    context: cwd,
    entry: entries.map((e) => `./${e}`),
    output: {
      path: path.resolve(cwd),
      filename: 'bundle.js',
    },
    target,
    externals: [nodeExternals()],
    module: {
      rules: [
        {
          test: /\.(graphql|tsx)$/,
          use: [
            {
              loader: 'babel-loader',
              options: {
                babelrc: false,
                configFile: false,
                presets: ['@babel/preset-react', '@babel/preset-typescript'],
              },
            },
            {
              loader: path.resolve(__dirname, '../../loader.ts'),
              options: gqlLetLoaderOptions,
            },
          ],
        },
      ],
    },
  });

  compiler.outputFileSystem = new memoryfs();

  return new Promise((resolve, reject) => {
    compiler.run((err, stats) => {
      if (err) reject(err);
      if (stats.hasErrors())
        reject(new Error(stats.toJson().errors.join('\n')));

      resolve(stats);
    });
  });
}
