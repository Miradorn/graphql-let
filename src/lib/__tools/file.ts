import copyfiles from 'copyfiles';
import { promises } from 'fs';
import { join as pathJoin, resolve as pathResolve, sep } from 'path';
import _rimraf from 'rimraf';
import { promisify } from 'util';

export const rimraf = promisify(_rimraf);

const crlf = /\r\n/g;
export function normalizeNewLine(str: string) {
  if (crlf.test(str)) return str.replace(crlf, '\n');
  return str;
}

export const { readFile: _readFile, writeFile, rename } = promises;

export function readFile(file: string) {
  return _readFile(file, 'utf-8').then(normalizeNewLine);
}

export function cleanup(cwd: string, relPaths: string[]) {
  return Promise.all(relPaths.map((rel) => rimraf(pathJoin(cwd, rel))));
}

function getDirDepth(fullDir: string) {
  let depth = 0;
  const dirs = fullDir.split(sep);
  for (let i = dirs.length - 1; i >= 0; i--) if (dirs[i].length) return i + 1;
  throw new Error('Never');
}

export function copyDir(
  baseFullDir: string,
  srcRelDir: string,
  destRelDir: string,
): Promise<void> {
  // baseFullDir: /Users/a/b/c
  // srcRelDir: d/e
  // destRelDir: .d/e
  // It generates /Users/a/b/c/.d/e
  if (srcRelDir === destRelDir) throw new Error('Kidding me?');
  const relPathOffset = 1;
  // strip empty dirs on sides
  const up =
    pathResolve(baseFullDir).split(sep).filter(Boolean).length + relPathOffset;
  const [relRoot] = destRelDir.split('/');
  return new Promise<void>((resolve, rejects) => {
    copyfiles(
      [pathJoin(baseFullDir, srcRelDir, '**'), pathJoin(baseFullDir, relRoot)],
      { error: true, up, all: true },
      (err) => {
        if (err) return rejects(err);
        resolve();
      },
    );
  });
}

export type AbsFn = (rel: string) => string;

export async function prepareFixtures(
  baseFullDir: string,
  fixtureSrcRelDir: string,
): Promise<[cwd: string, abs: AbsFn]> {
  if (fixtureSrcRelDir.startsWith('..'))
    throw new Error(
      `It doesn't support backward relative paths like ${fixtureSrcRelDir}`,
    );
  const fixtureDestRelDir = '.' + fixtureSrcRelDir;
  const cwd = pathJoin(baseFullDir, fixtureDestRelDir);
  const abs = (relPath: string) => pathJoin(cwd, relPath);
  await copyDir(baseFullDir, fixtureSrcRelDir, fixtureDestRelDir);
  return [cwd, abs];
}
