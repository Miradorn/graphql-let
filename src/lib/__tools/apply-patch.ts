import { applyPatch as diffApplyPatch } from 'diff';
import { join } from 'path';
import { readFile, writeFile } from './file';

export async function applyPatch(
  cwd: string,
  targetRelPath: string,
  patchRelPath: string,
) {
  await writeFile(
    join(cwd, targetRelPath),
    diffApplyPatch(
      await readFile(join(cwd, targetRelPath)),
      await readFile(join(cwd, patchRelPath)),
    ),
  );
}
