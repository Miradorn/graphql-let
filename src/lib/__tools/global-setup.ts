import { join } from 'path';
import { cleanup, prepareFixtures } from './file';

const projRoot = join(__dirname, '../../..');

export default async function setup() {
  await cleanup(projRoot, ['**/.__fixtures']);

  // XXX: babel-plugin-tester wants this.
  await prepareFixtures(
    join(__dirname, '../../../__tests__'),
    '__fixtures/babel',
  );
}
