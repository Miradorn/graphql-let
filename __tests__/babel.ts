import pluginTester, { TestObject } from 'babel-plugin-tester';
import { join } from 'path';
import myPlugin from '../src/babel-plugin';
import { spawn } from '../src/lib/__tools/child-process';

const fixtureBaseDir = join(__dirname, '.' + '__fixtures/babel/suite');

const tests: TestObject[] = ['macro-gql', 'macro-load', 'plugin'].map(
  (name) => {
    const cwd = join(fixtureBaseDir, name);
    return {
      title: name,
      babelOptions: { cwd },
      teardown: async () => {
        await spawn('yarn', ['tsc'], { cwd });
      },
      fixture: join(fixtureBaseDir, name, 'input.ts'),
    };
  },
);

pluginTester({
  title: 'graphql-let',
  plugin: myPlugin,
  snapshot: true,
  babelOptions: {
    plugins: [
      'graphql-let/babel',
      [
        'babel-plugin-module-resolver',
        {
          alias: { 'graphql-let': './' }, // Needed only for tests.
        },
      ],
      'babel-plugin-macros',
      '@babel/plugin-transform-typescript',
    ],
  },
  tests,
});

// const ignoreDirs: string[] = [
//   // 'macro-load'
// ];
//
// beforeAll(async () => {
//   await cleanup(cwd, ['**/node_modules']);
// });
//
// runner(
//   cwd,
//   'gql',
//   // {},
//   {
//     ignoreTasks: ignoreDirs.map((d) => d.split('-').join(' ')),
//   },
//   { sourceType: 'unambiguous' },
// );
//
// test(`Type checking for all fixtures`, async () => {
//   const dirs = await glob(['*/*', ...ignoreDirs.map((d) => `!*/${d}`)], {
//     cwd,
//     onlyDirectories: true,
//     absolute: true,
//   });
//   for (const dir of dirs) await spawn('yarn', ['tsc'], { cwd: dir });
// });
//
// test(`Generated files are okay`, async () => {
//   await matchPathsAndContents(['**/node_modules/**'], cwd);
// });
