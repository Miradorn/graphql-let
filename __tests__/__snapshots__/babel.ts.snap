// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphql-let macro-gql: macro-gql 1`] = `

import { gql } from 'graphql-let/macro';

const { useViewerQuery } = gql(\`
  # import Partial from './documents/partial.graphql'
  query Viewer {
    viewer {
      ...Partial
    }
  }
\`);

function nest1() {
  function nest2() {
    function nest3() {
      function nest4() {
        function MyComponent() {
          useViewerQuery().data.viewer.name as string;
          // @ts-expect-error
          useViewerQuery() as number;
        }
      }
    }
  }
}

      ↓ ↓ ↓ ↓ ↓ ↓

import * as Va8ca7f2176b9dce4e4ba81526fcae512e2279848 from './node_modules/graphql-let/__generated__/input-Viewer-Partial.ts';
const { useViewerQuery } = Va8ca7f2176b9dce4e4ba81526fcae512e2279848;

function nest1() {
  function nest2() {
    function nest3() {
      function nest4() {
        function MyComponent() {
          useViewerQuery().data.viewer.name; // @ts-expect-error

          useViewerQuery();
        }
      }
    }
  }
}


`;

exports[`graphql-let macro-load: macro-load 1`] = `

import { load } from 'graphql-let/macro';

const { useViewerQuery } = load(\`./documents/viewer.graphql\`);

function nest1() {
  function nest2() {
    function nest3() {
      function nest4() {
        function MyComponent() {
          useViewerQuery().data.viewer.name as string;
          // @ts-expect-error
          useViewerQuery() as number;
        }
      }
    }
  }
}

      ↓ ↓ ↓ ↓ ↓ ↓

import * as Va8ca7f2176b9dce4e4ba81526fcae512e2279848 from './node_modules/graphql-let/__generated__/input-Viewer-Partial.ts';
const { useViewerQuery } = Va8ca7f2176b9dce4e4ba81526fcae512e2279848;

function nest1() {
  function nest2() {
    function nest3() {
      function nest4() {
        function MyComponent() {
          useViewerQuery().data.viewer.name; // @ts-expect-error

          useViewerQuery();
        }
      }
    }
  }
}


`;

exports[`graphql-let plugin: plugin 1`] = `

import { gql } from 'graphql-let';

const { useViewerQuery } = gql(\`query Viewer { viewer { name } }\`);

function nest1() {
  function nest2() {
    function nest3() {
      function nest4() {
        function MyComponent() {
          useViewerQuery().data.viewer.name as string;
          // @ts-expect-error
          useViewerQuery() as number;
        }
      }
    }
  }
}

      ↓ ↓ ↓ ↓ ↓ ↓

import * as V4de3a990eddfa70dba05f32dbaca0afc6aca7bda from './node_modules/graphql-let/__generated__/input-Viewer.ts';
const { useViewerQuery } = V4de3a990eddfa70dba05f32dbaca0afc6aca7bda;

function nest1() {
  function nest2() {
    function nest3() {
      function nest4() {
        function MyComponent() {
          useViewerQuery().data.viewer.name; // @ts-expect-error

          useViewerQuery();
        }
      }
    }
  }
}


`;
